### PR TITLE
Redis Cluster 3.0 configuration example

### PR DESCRIPTION
**Pull request by balonus**
Example of reliable 6 nodes Redis Cluster (3 masters + 3 slaves) - based on [Redis Cluster 3.0](http://redis.io/topics/cluster-spec) functionality. This is different approach then [Redis Sentinel](http://redis.io/topics/sentinel) presented in https://github.com/kubernetes/kubernetes/tree/master/examples/redis

---

This change is [Reviewable](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/22597)

### CODE REVIEW

**Comment by rasekar2000** [comment]
Team, Is there anyway to see this source code?

**Comment by rasekar2000** [comment]
@balonus Is this redis-cluster , kubernetes example merged into master branch? Please let me know. Thanks.

**Comment by balonus** [comment]
Hi @rasekar2000 ,

No it is not merged into master.
It's available here: 
https://github.com/balonus/kubernetes/tree/redis-cluster-example/examples/redis-cluster

**Comment by bgrant0607** [comment]
cc @kubernetes/examples @bprashanth

**Comment by bgrant0607** [comment]
@balonus Is there a reason we should keep the existing redis example?

**Comment by bgrant0607** [comment]
If you haven't already, PTAL at:
https://github.com/kubernetes/kubernetes/blob/master/examples/guidelines.md

**Comment by bprashanth** [comment]
@balonus just fyi, we've got a couple of things in flight that will make this easier. Take a look at the rough redis implementation with petset: https://github.com/kubernetes/contrib/pull/888/commits/31387e37ccbd2b67b06974f868dbb4131bc0368f, it only spins up a 3 pod cluster in master/slave, but I think it captures the essence of the idea.

I'd suggest using hostnames instead of ips. The petset will give you a predictable start order, and predictable hostnames. In addition, there's a "peer-finder" helper that sends on-start and on-change notification events (https://github.com/kubernetes/contrib/pull/921/commits/b860ab915cb7b87b647fe9919f8e6d4f69ded393). You can utilize this to write out your config.

You can actually clean up even further with init containers, like in the zookeeper example: https://github.com/kubernetes/contrib/pull/921, but that depends on (as of this writing) unmerged code.

Take a look at these pointers and let me know if you have any questions. I'm fine tackling this iteratively as well (i.e follow up prs that converge to the petset idea), especially since the petset controller is currently only in HEAD.

**Comment by balonus** [comment]
@bgrant0607 
> Is there a reason we should keep the existing redis example?

Existing redis example demonstrates Redis Sentinel approach while my proposal demonstrates Redis Cluster. Both solutions are alternate ways to achieve Redis HA (both are redis native mechanisms). So I think it's worth to keep them both in examples.

@bprashanth thanks for pointing to your example with petset & peer-finder - I will take a look into it.
I need to review my example once more and want to upgrade it as new Redis 3.2 has just been released.

**Comment by chrislovecnm** [comment]
@balonus I read 3.2 includes some new proxying features for docker.  Also the example in the main project should eventually have an example for PetSet ;)

We have the same challenges with PetSet when failover occurs and slave / master election happens.

But with sentinel it is pretty clean. It would be good to migrate all of this to under examples ;)

**Comment by chrislovecnm** [comment]
Can we ge a makefile to deploy the containers to google example repo?

**Comment by bprashanth** [comment]
> We have the same challenges with PetSet when failover occurs and slave / master election happens.

@chrislovecnm this is probably out of scope of petset, the challenges here should be the same challenges one faces with vms. Down the line we might have different "types" of petset, but at this point it's not clear that would be any benefit, given things like sentinel and mysql utilities for failover already exist. 

Petset will not, for example, elect your master. You can take a key in some k/v store and use the usual ttl trick to notify slaves that the master is down, or deploy something like sentinel if your application supports it, or use the election library (https://github.com/kubernetes/contrib/blob/d9063241280cedfaac23a2283f6584e747aa2302/election/README.md) and write your own babysitter.

**Comment by chrislovecnm** [comment]
I would never expect k8s to handle the elections. That is the job of sentinel. 

Let each datastore manage its own cluster, would be the rule that I would follow. PetSet gives us persistent disk that we can drop rdb to in redis, fixed ip for sentinels, etc. Much like C*.  With redis there are a lot of different deployment patterns, unlike  c*, but my hallucination is that they will work with PetSet.

**Comment by bgrant0607** [comment]
cc @smarterclayton

File: examples/redis-cluster/README.md [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+## Reliable Redis Cluster 3.0 on Kubernetes
+
+The following document describes the deployment of a reliable Redis 3.0 Cluster on Kubernetes. It deploys 6 redis nodes to form redis cluster consisting of 3 masters and 3 slaves.
+
+### Prerequisites
+
+This example assumes that you have a Kubernetes cluster installed and running, and that you have installed the ```kubectl``` command line tool somewhere in your path.  Please see the [getting started](../../docs/getting-started-guides/) for installation instructions for your platform.
+
+Presented here approach works only when Kubernetes cluster operates in [iptables](https://github.com/kubernetes/kubernetes/issues/3760) mode (eg. by running `kube-proxy` with option `--proxy-mode=iptables`).
+When working with [Google Container Engine](https://cloud.google.com/container-engine/) this mode can be set by running:
```
Comment (viglesiasce) [inline examples/redis-cluster/README.md]: This is no longer needed as iptables is now the default on GKE.

File: examples/redis-cluster/README.md [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+replicationcontroller "redis-001-node-201" created
+replicationcontroller "redis-001-node-202" created
+replicationcontroller "redis-001-node-301" created
+replicationcontroller "redis-001-node-302" created
+```
+
+###### Check redis cluster topology & status
+
+```sh
+examples/redis-cluster/scripts/redis-cluster-info.sh
```
Comment (viglesiasce) [inline examples/redis-cluster/README.md]: The script isn't marked executable so this needs to change to:

```sh
sh examples/redis-cluster/scripts/redis-cluster-info.sh
```

**Comment by viglesiasce** [comment]
In the case of 1 service and 1 pod (rc of replica 1) per Redis node, which service should external clients be connecting to? 

It seems like there needs to be something that can decide when nodes are healthy and ready to receive traffic. For example, if I am hitting redis-001-node-101.default.svc from within the cluster and that node goes down for some reason, the outage will be perceptible by the client until that pod comes back up somewhere else.

**Comment by rasekar2000** [comment]
Yes. Using Redis Sentinel based clustering now. Want to explore Redis native cluster in Kubernetes as soon as its ready. Thanks

**Pull request closed without merging**